contracts/debt/IncurDebtV1.sol

```diff
@@ -0,0 +1,458 @@
+error IncurDebtV1_BorrowerHasNoOutstandingDebt(address _borrower);
+error IncurDebtV1_BorrowerStillHasOutstandingDebt(address _borrower);
+
+contract IncurDebtV1 is OlympusAccessControlledV2, IIncurDebtV1 {
+    using SafeERC20 for IERC20;
+
+    event GlobalLimitChanged(uint256 indexed _limit);
+    event BorrowerAllowed(address indexed _borrower);
+    event BorrowerRevoked(address indexed _borrower);
+    event BorrowerDebtLimitSet(address indexed _borrower, uint256 indexed _limit);
+    event CollateralInSOHMIncreased(address indexed _borrower, uint256 indexed _currentCollateralInSOHM);
+    event BorrowerDeposit(
+        address indexed _borrower,
+        uint256 indexed _amountToDeposit,
+        address indexed _tokenToDepositFundsWith
+    );
+    event Borrowed(
+        address indexed _borrower,
+        uint256 _amountToBorrow,
+        uint256 indexed _borrowersDebt,
+        uint256 indexed _totalOutstandingGlobalDebt
+    );
+    event DebtPaidWithOHM(
+        address indexed _borrower,
+        uint256 _paidDebt,
+        uint256 indexed _currentDebt,
+        uint256 indexed _totalOutstandingGlobalDebt
+    );
+    event Withdrawal(
+        address indexed _borrower,
+        uint256 _amountToWithdraw,
+        address _tokenToWithdrawFundsWith,
+        address indexed _receiver,
+        uint256 indexed _collateralInSOHMLeft
+    );
+    event DebtPaidWithCollateral(
+        address indexed _borrower,
+        uint256 indexed _paidDebt,
+        uint256 _currentCollateralInSOHM,
+        uint256 indexed _currentDebt,
+        uint256 _totalOutstandingGlobalDebt
+    );
+    event DebtPaidWithCollateralAndBurnTheRest(
+        address indexed _borrower,
+        uint256 indexed _paidDebt,
+        uint256 _currentCollateralInSOHM,
+        uint256 _currentDebt,
+        uint256 _totalOutstandingGlobalDebt,
+        uint256 indexed _collateralLeftToBurn
+    );
+    event DebtPaidWithCollateralAndWithdrawTheRest(
+        address indexed _borrower,
+        uint256 indexed _paidDebt,
+        uint256 _currentCollateralInSOHM,
+        uint256 _currentDebt,
+        uint256 _totalOutstandingGlobalDebt,
+        uint256 indexed _collateralLeftForWithdraw
+    );
+    event ForceDebtPayWithCollateralAndWithdrawTheRest(
+        address indexed _borrower,
+        uint256 indexed _paidDebt,
+        uint256 _currentCollateralInSOHM,
+        uint256 _currentDebt,
+        uint256 _totalOutstandingGlobalDebt,
+        uint256 indexed _collateralLeftForWithdraw
+    );
+
+    uint256 public globalDebtLimit;
+    uint256 public totalOutstandingGlobalDebt;
+
+    address public immutable OHM;
+    address public immutable gOHM;
+    address public immutable sOHM;
+    address public immutable staking;
+    address public immutable treasury;
+
+    struct Borrower {
+        uint128 debt;
+        uint128 limit;
+        uint128 collateralInSOHM;
+        uint128 collateralInGOHM;
```

Comment:
Should call this somehting like `indexedCollateral` or something similar. DM me if you have questions.